index.js

```diff
@@ -49,122 +74,221 @@ function retrieve(program, callback){
+                if(err) return recordCallback(err);
+                count--;
+                if(entriesFinished && !count){
+                  var unzipped = path.join(zipdir, record.file);
+
+                  if(csvReg.test(record.file)){
+
+                    csvToVrt(unzipped, record.sourceSrs, function(err, vrt){
+                      if(err) return recordCallback(err);
+                      handleStream(spawnOgr(vrt), record);
+                    });
+
+                  }else{
+                    handleStream(spawnOgr(unzipped), record);
+                  }
+                }
+              });
             });
-          }else{
-            handleStream(spawnOgr(unzipped), record, recordCallback);
-          }
+          });
         });
+      });
     }else{
       if(csvReg.test(record.file)){
         var csv = path.join(scratchSpace, record.file);
         var csvStream = fs.createWriteStream(csv);
 
-        csvStream.on('finish', function(err){
+        pump(stream, csvStream, function(err){
           if(err) return recordCallback(err);
 
           csvToVrt(csv, record.sourceSrs, function(err, vrt){
             if(err) return recordCallback(err);
-
-            handleStream(spawnOgr(vrt), record, recordCallback);
+            handleStream(spawnOgr(vrt), record);
           });
         });
 
-        request.pipe(csvStream);
       }else{
-        handleStream(spawnOgr(null, request), record, recordCallback);
+        handleStream(spawnOgr(null, stream), record);
       }
     }
-  });
+  }
+
+
+  function handleStreamError(err){
+    if(this.unpipe) this.unpipe();
+    if(this.destroy) this.destroy();
+    if(this.kill) this.kill();
+    recordCallback(err);
+  }
 
 
   function spawnOgr(file, stream){
-    var child; 
+    var jsonChild;
+
     if(stream){
-      child = spawn('ogr2ogr', ['-f', 'CSV', '-t_srs', 'WGS84', '-lco', 'GEOMETRY=AS_XY', '/vsistdout/', '/vsistdin/'])
-      stream.pipe(child.stdin);
+      jsonChild = spawn('ogr2ogr', ['-f', 'GeoJSON', '-t_srs', 'WGS84', '/vsistdout/', '/vsistdin/']);
+      pump(stream, jsonChild.stdin);
     }else{
-      child = spawn('ogr2ogr', ['-f', 'CSV', '-t_srs', 'WGS84', '-lco', 'GEOMETRY=AS_XY', '/vsistdout/', file])
+      jsonChild = spawn('ogr2ogr', ['-f', 'GeoJSON', '-t_srs', 'WGS84', '/vsistdout/', file]);
     }
-    return child.stdout;
+
+    var csvChild = spawn('ogr2ogr', ['-f', 'CSV', '-lco', 'GEOMETRY=AS_XY', '/vsistdout/', '/vsistdin/']);
+    var ogrJsonStream = OgrJsonStream();
+    var centroids = centroidStream.stringify();
+
+    jsonChild.stderr.on('data', function(errorText){
+      jsonChild.stdout.emit('error', errorText);
+    });
+
+    csvChild.stderr.on('data', function(errorText){
+      csvChild.stdout.emit('error', errorText);
+    });
+
+    pump(jsonChild.stdout, ogrJsonStream, centroids, function(){
+      csvChild.stdin.end(']}');
+    });
+
+    csvChild.stdin.write('{"type":"FeatureCollection","features":[');
```

Comment:
This seems kinda gross (building JSON via string concat), but I'm guessing that's our only real option when spawning a call to `ogr2ogr`.  It _seems_ safe since its just concating JSON snippets into a wrapping JSON object.  Just want to make sure we're taking into consideration any edge cases where `ogr2ogr` may spit out something other than JSON when something goes awry.  So long as `ogr2ogr` always writes to stderr when something goes wrong, this seems like we're in good shape?